Remove FailedResourceType and return custom error

hongchaodeng opened the pull request:
Currently in Fit predicate it uses a global variable `FailedResourceType` to pass information:

``` Go
            predicates.FailedResourceType = ""
            fit, err := predicate(pod, machineToPods[node.Name], node.Name)
            if !fit {
                ...
                if predicates.FailedResourceType != "" {
                    failedPredicateMap[node.Name].Insert(predicates.FailedResourceType)
                    break
                }
            }
```

We should use custom error to pass it instead:

```
            fit, err := predicate(pod, machineToPods[node.Name], node.Name)
            if err != nil {
                return api.NodeList{}, FailedPredicateMap{}, err
            }
            if !fit {
                ...
                if re, ok := err.(*predicates.ErrResourceUnfit); ok {
                    failedPredicateMap[node.Name].Insert(re.Reason())
                    break
                }
            }
```

This PR adds a new custom type of error to predicates

xiang90 commented:
@hongchaodeng Thanks! I wanted to remove this a while ago.

hongchaodeng commented:
@lavalamp 
Can you take a look at this PR?

lavalamp commented:
If I understand what you're doing with this--please make a custom error type and encode this information in the error. Don't add an additional return parameter for this.

hongchaodeng committed:
Remove FailedResourceType and return custom error

hongchaodeng commented:
@lavalamp 
Thanks for the suggestion. Addressed.

xiang90 commented on plugin/pkg/scheduler/algorithm/predicates/error.go:
```diff
@@ -0,0 +1,19 @@
+
+import "fmt"
+
+type ResourceUnFitError struct {
```
type ErrResourceUnfit?

hongchaodeng replied:
Fixed

hongchaodeng commented:
@gmarek 
Can you help review the PR?

lavalamp commented on plugin/pkg/scheduler/algorithm/predicates/error.go:
```diff
@@ -0,0 +1,44 @@
+
+// ErrResourceUnfit is an error type that indicates what kind of resource limit is
+// hit and caused the unfitting failure.
+type ErrResourceUnfit struct {
```
InsufficientResourceError

lavalamp commented on plugin/pkg/scheduler/algorithm/predicates/error.go:
```diff
@@ -0,0 +1,44 @@
+// ErrResourceUnfit is an error type that indicates what kind of resource limit is
+// hit and caused the unfitting failure.
+type ErrResourceUnfit struct {
+	reason string
```
Don't call this reason if it's actually just the resource name.

lavalamp replied:
You could consider adding requested & available fields.

lavalamp replied:
If you expect the error to escape the package, it's actually a good idea to make the internal fields public so that users can get at them without parsing the error message.

lavalamp commented on plugin/pkg/scheduler/algorithm/predicates/error.go:
```diff
@@ -0,0 +1,44 @@
+var (
+	ErrPodExceedsMaxPodNumber = newResourceUnfitError("PodExceedsMaxPodNumber")
+	ErrPodExceedsFreeCPU      = newResourceUnfitError("PodExceedsFreeCPU")
+	ErrPodExceedsFreeMemory   = newResourceUnfitError("PodExceedsFreeMemory")
```
Given the message formatting you used, the "PodExceeds" part of this is not necessary.

MaxPods
CPU
Memory

lavalamp replied:
So if you want to just have static errors and not parameterize them at all (i.e., you don't want to put in requested/available etc)-- then you could just use errors.New() to create all of these.

Generally, if you make an error type then I expect it to be parameterizable and I expect that direct comparison is not the proper way to check it. (A type assertion + looking at its field(s) would be the proper way.)

I think given what you want to do, using errors.New() is less work.

hongchaodeng replied:
I didn't catch what you mean. Can you be more explicit?
Let me give explicit reasons for the changes:
- We need custom error type instead of errors.New() in order to differentiate resource unfit cases
- Those constants are shared across different packages and tests. I'm not sure what "parameterizable" means here? Is it orthogonal?

hongchaodeng commented:
@lavalamp Address the comments. PTAL and feel free to give any feedback if it still isn't what you mentioned.

lavalamp commented on plugin/pkg/scheduler/algorithm/predicates/error.go:
```diff
@@ -0,0 +1,39 @@
+}
+
+func (e *InsufficientResourceError) Error() string {
+	return fmt.Sprintf("InsufficientResourceError: %s", e.FailedResourceType)
```
I think your previous text was better :)

lavalamp commented on plugin/pkg/scheduler/algorithm/predicates/error.go:
```diff
@@ -0,0 +1,39 @@
+	FailedResourceType string
+}
+
+func newResourceUnfitError(failedResourceType string) *InsufficientResourceError {
```
newInsufficientResourceError

lavalamp commented on plugin/pkg/scheduler/algorithm/predicates/error.go:
```diff
@@ -0,0 +1,39 @@
+// ErrResourceUnfit is an error type that indicates what kind of resource limit is
+// hit and caused the unfitting failure.
+type InsufficientResourceError struct {
+	FailedResourceType string
```
I would call this just "Resource" or "ResourceName".

lavalamp commented on plugin/pkg/scheduler/algorithm/predicates/error.go:
```diff
@@ -0,0 +1,39 @@
+var (
+	ErrExceededMaxPodNumber   = newResourceUnfitError("ExceededMaxPodNumber")
+	ErrInsufficientFreeCPU    = newResourceUnfitError("InsufficientFreeCPU")
+	ErrInsufficientFreeMemory = newResourceUnfitError("InsufficientFreeMemory")
```
The below makes more sense to me.

```
 	ErrExceededMaxPodNumber   = newResourceUnfitError("PodCount")
 	ErrInsufficientFreeCPU    = newResourceUnfitError("CPU")
	ErrInsufficientFreeMemory = newResourceUnfitError("Memory")
```


I was trying to say, given your usage (only one parameter), the below will work for you, too:

```
 	ErrExceededMaxPodNumber   = errors.New("Insufficient resources: PodCount")
 	ErrInsufficientFreeCPU    = errors.New("Insufficient resources: CPU")
	ErrInsufficientFreeMemory = errors.New("Insufficient resources: Memory")
```

You *can* compare errors, i.e., `err == ErrExceededMaxPodNumber` if you do this, and it will work like you want.

hongchaodeng commented:
@lavalamp 
Addressed comments.
I see your point. But I still prefer to use custom error because it's easier to check and differentiate.
```
				if re, ok := err.(*predicates.InsufficientResourceError); ok {
					failedPredicateMap[node.Name].Insert(re.ResourceName)
					break
				}
```

PTAL and let me know what you think again. Thanks!

lavalamp commented:
LGTM, thanks

mwielgus commented:
k8s bot test this please

davidopp commented:
@k8s-bot unit test this please

gmarek commented:
@k8s-bot unit test this

gmarek commented:
I believe this broke our tests:
```
SchedulerPredicates [Serial] validates resource limits of pods that are allowed to run [Conformance]

/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/scheduler_predicates.go:355 Expected : 19 to equal : 21
```

I'll try reverting it locally to confirm.

hongchaodeng commented:
@gmarek 
I'm trying to understand what error it brings. But the error message is so cryptic (who brought up the idea to use ginkgo?) My guess is that this line failed: https://github.com/kubernetes/kubernetes/blob/14d9a0f2c878f249e973a83a214bf077e14f5949/test/e2e/scheduler_predicates.go#L86

If so, it looks like there is terrible logical error in listing pods..

yujuhong commented:
Can we revert the PR to unblock the builds if the test/pr cannot be fixed immediately?

hongchaodeng commented:
Can we wait for @gmarek 's confirmation?

saad-ali commented:
I too suspect that https://github.com/kubernetes/kubernetes/issues/19487 is caused by this PR. This is blocking the merge queue and is a P0 issue, so I will roll back this PR (it can be re-merged if it is not problematic).

gmarek commented:
After revert it was still failing. Sorry - I had to run home:/

gmarek commented:
It's better if @saad-ali would take it from here.

saad-ali commented:
Thanks @gmarek. I'll take over.

saad-ali commented:
Ok, I'll hold off on rolling back this PR then. I'll look into some of the other PRs that went in at the same time, if nothing looks promising I'll roll back anyway.

saad-ali commented:
This PR still looks the most suspect to me. I'm going to roll it back to be safe. If it is not the cause of the issue we can re-merge when everything is green again.

saad-ali commented:
Confirmed revert of this PR fixed #19487
